File: java_refactoring/src/main/java/org/example2/solution/TeaClient.java
package org.example2.solution;

import java.util.List;
import java.util.function.Supplier;

public class TeaClient {
    private static final List<TeaOrder> TEA_ORDERS = List.of(
            new TeaOrder(() -> new GreenTeaService(120), new Cup("mug", "white", "ceramic", 250, 300)),
            new TeaOrder(() -> new BlackTeaService(100), new Cup("cup", "blue", "glass", 200, 250)),
            new TeaOrder(() -> new HerbalTeaService(122), new Cup("teapot", "red", "metal", 500, 600))
    );

    public static void main(String[] args) {
        TeaClient client = new TeaClient();
        client.prepareOrders();
    }

    public void prepareOrders() {
        for (TeaOrder order : TEA_ORDERS) {
            TeaService tea = order.createTea();
            tea.makeTea();
            tea.pourTea(order.cup());
            payTaxes(tea.createTaxInfo());
        }
    }

    private void payTaxes(TeaTaxInfo taxInfo) {
        System.out.println("Paying total tax: " + taxInfo.getTotalTax() + " property tax: " + taxInfo.getPropertyTax());
        System.out.println("Paid total tax: " + taxInfo.getTotalTax() + " property tax: " + taxInfo.getPropertyTax());
    }

    private record TeaOrder(Supplier<TeaService> teaFactory, Cup cup) {
        private TeaService createTea() {
            return teaFactory.get();
        }
    }
}

File: typescript_refactoring/example2/solution/TeaClient.ts
import { BlackTeaService } from "./BlackTeaService";
import { GreenTeaService } from "./GreenTeaService";
import { HerbalTeaService } from "./HerbalTeaService";
import { TeaService } from "./TeaService";
import { Cup, TeaTaxInfo } from "./TeaTypes";

const teaOrders: Array<{ createTea: () => TeaService; cup: Cup }> = [
    {
        createTea: () => new GreenTeaService(120),
        cup: { type: "mug", color: "white", material: "ceramic", sizeMl: 250, weightGrams: 300 },
    },
    {
        createTea: () => new BlackTeaService(100),
        cup: { type: "cup", color: "blue", material: "glass", sizeMl: 200, weightGrams: 250 },
    },
    {
        createTea: () => new HerbalTeaService(122),
        cup: { type: "teapot", color: "red", material: "metal", sizeMl: 500, weightGrams: 600 },
    },
];

export class TeaClient {
    public static main(): void {
        const client = new TeaClient();
        client.prepareOrders();
    }

    public prepareOrders(): void {
        for (const order of teaOrders) {
            const tea = order.createTea();
            tea.makeTea();
            tea.pourTea(order.cup);
            this.payTaxes(tea.createTaxInfo());
        }
    }

    private payTaxes(taxInfo: TeaTaxInfo): void {
        console.log(`Paying total tax: ${taxInfo.totalTax} property tax: ${taxInfo.propertyTax}`);
        console.log(`Paid total tax: ${taxInfo.totalTax} property tax: ${taxInfo.propertyTax}`);
    }
}
